### ISSUE TITLE
3D pynufft is crashing over >(64,64,64) volumes

### ISSUE THREAD

**Issue by bensarthou** [jyhmiinlin/pynufft#10]
When I try to call Pynufft (CPU-only) over a 3D volume more than (64,64,64), the program crashes over plan() (more specifically precompute_sp) with a MemoryError (see traceback below) 

< errors occur in self.precompute_sp()
Traceback (most recent call last):
  File "test_pynufft.py", line 121, in <module>
    NufftObj.plan(om, Nd, Kd, Jd)
  File "/home/bs255482/.local/lib/python3.5/site-packages/pynufft/src/_transform/transform_cpu.py", line 74, in plan
    self._precompute_sp()
  File "/home/bs255482/.local/lib/python3.5/site-packages/pynufft/src/_transform/transform_cpu.py", line 96, in _precompute_sp
    self.spHsp =self.st['p'].getH().dot(self.st['p']).tocsr()
  File "/home/bs255482/.local/lib/python3.5/site-packages/scipy-1.0.0-py3.5-linux-x86_64.egg/scipy/sparse/base.py", line 322, in dot
    return self * other
  File "/home/bs255482/.local/lib/python3.5/site-packages/scipy-1.0.0-py3.5-linux-x86_64.egg/scipy/sparse/base.py", line 440, in __mul__
    return self._mul_sparse_matrix(other)
  File "/home/bs255482/.local/lib/python3.5/site-packages/scipy-1.0.0-py3.5-linux-x86_64.egg/scipy/sparse/compressed.py", line 503, in _mul_sparse_matrix
    data = np.empty(nnz, dtype=upcast(self.dtype, other.dtype))
MemoryError>

My machine is equipped with 32 Intel Xeon Core, which is normally enough to deal with large datas. (the same transform performs well with pynfft) 

I wonder if you have seen this limitation before and if so, maybe a warning in the module could be useful for other users. 
(P.S I'm the intern working with P. Ciuciu, I have send you a mail about others issues I found in Pynufft)

**Comment by jyhmiinlin** [comment]
Sorry for my delayed response. I had been working on reducing the memory size. I am thinking the forward gridding matrix can be reused in the adjoint operation.

**Comment by jyhmiinlin** [comment]
Solved in 2019.1.2. Closed.

The detail is described in https://arxiv.org/abs/1903.08365.

In short, each non-Cartesian point in 3D MRI requires a "cube-like" interpolator. Each cube is a rank-1 tensor. Therefore, it can be represented as three vectors. Three vectors are smaller than a cube. 
So I rewrote the cSpmv and cSpmvh to use vectors and the tensor is constructed at run-time. 

Thanks to Benoit, Loubna, Dr. Ciuciu for raising this issue and testing the 3D MRI.
